calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java

```diff
@@ -0,0 +1,186 @@
+package com.market.banica.calculator.service;
+
+import com.market.banica.calculator.data.contract.RecipesBase;
+import com.market.banica.calculator.dto.ProductDto;
+import com.market.banica.calculator.model.Product;
+import com.market.banica.calculator.service.contract.BackUpService;
+import com.market.banica.calculator.service.contract.ProductService;
+import io.micrometer.core.lang.Nullable;
+import lombok.RequiredArgsConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class ProductServiceImpl implements ProductService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProductServiceImpl.class);
+
+    private final RecipesBase recipesBase;
+    private final BackUpService backUpService;
+
+    @Override
+    public Product createProduct(List<Product> products) {
+        LOGGER.debug("Product service impl: In createProduct method");
+
+        validateParameterForNullAndEmpty(products);
+
+        validateAllProductsInListAreNew(products);
+
+        createProductsInDatabase(products);
+
+        createBackUp();
+
+        String recipeName = getRecipeName(products);
+
+        LOGGER.debug("Recipe {} successfully created", recipeName);
+        return recipesBase.getDatabase().get(recipeName);
+    }
+
+    @Override
+    public List<ProductDto> getProduct(String productName, @Nullable String parentProductName) {
+        LOGGER.debug("Product service impl: In getProduct method");
+
+        Product product = getProductFromDatabase(productName);
+
+        validateProductExist(productName, product);
```

Comment:
This validation should happen inside the getProductFromDatabase method

Reply:
Done